runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/OpAdapter.java

```diff
@@ -188,7 +176,31 @@ public void emitWatermark(Instant watermark) {
 
     @Override
     public <T> void emitView(String id, WindowedValue<Iterable<T>> elements) {
-      outputList.add(OpMessage.ofSideInput(id, elements));
+      outputQueue.add(OpMessage.ofSideInput(id, elements));
+    }
+
+    @Override
+    public Collection<OpMessage<OutT>> collectOutput() {
+      final List<OpMessage<OutT>> outputList = new ArrayList<>();
+      OpMessage<OutT> output;
+      while ((output = outputQueue.poll()) != null) {
+        outputList.add(output);
+      }
+      return outputList;
+    }
+
+    @Override
+    public CompletionStage<Collection<OpMessage<OutT>>> collectFuture() {
+      final CompletionStage<Collection<OpMessage<OutT>>> future = outputFuture;
+      outputFuture = null;
```

Comment:
this needs to be synchronized across `emitFuture(...)`?

Reply:
The caller apply() is synchronized so this it thread-safe.

Reply:
if you mean collectOutput() is also invoked by AsyncDoFnRunner, I already replaced the underlying data structure to be ConcurrentLinkedQueue so the impl is thread-safe. For collectFuture(), it's only used by OpAdapter.apply(), and the caller is thread-safe.

Reply:
Sounds good.